staging/src/k8s.io/client-go/transport/cert_rotation.go

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package transport
+
+import (
+	"crypto/tls"
+	"fmt"
+	"reflect"
+	"sync/atomic"
+	"time"
+
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/connrotation"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+)
+
+const workItemKey = "key"
+
+// CertCallbackRefreshDuration is exposed so that integration tests can crank up the reload speed.
+var CertCallbackRefreshDuration = 5 * time.Minute
+
+type rotatingClientCert struct {
+	clientCert *atomic.Value
+	callback   func(*tls.CertificateRequestInfo) (*tls.Certificate, error)
+	connDialer *connrotation.Dialer
+
+	// queue only ever has one item, but it has nice error handling backoff/retry semantics
+	queue workqueue.RateLimitingInterface
+}
+
+func certRotatingDialer(tlsConfig *tls.Config, dial utilnet.DialFunc) utilnet.DialFunc {
+	d := &rotatingClientCert{
+		clientCert: new(atomic.Value),
+		callback:   tlsConfig.GetClientCertificate,
+		connDialer: connrotation.NewDialer(connrotation.DialFunc(dial)),
+		queue:      workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "DynamicClientCertificate"),
+	}
+
+	go d.run(1, make(chan struct{}))
+
+	tlsConfig.GetClientCertificate = d.GetClientCertificate
+
+	return d.connDialer.DialContext
+}
+
+// loadClientCert calls the callback and rotates connections if needed
+func (c *rotatingClientCert) loadClientCert() (*tls.Certificate, error) {
+	cert, err := c.callback(nil)
+	if err != nil {
+		return cert, err
+	}
+
+	// check to see if we have a change. If the values are the same, do nothing.
+	existing, ok := c.clientCert.Load().(*tls.Certificate)
+	if ok && existing != nil && reflect.DeepEqual(existing, cert) {
```

Comment:
Will cert.Leaf every be lazily filled in? I think that would break this check.

Reply:
This is how we do it in the exec auth provider. Will write a check excluding the leaf